File: Java/src/com/dbc/algorithm/Array/LongestSubstringWithoutRepeatingCharacters.java
package com.dbc.algorithm.Array;

import java.util.HashMap;
import java.util.Map;

public class LongestSubstringWithoutRepeatingCharacters {
    public static int lengthOfLongestSubstring(String s) {
        int result = 0, left = 0;
        Map<Character, Integer> count = new HashMap<>();
        for (int i = 0; i < s.length(); i++) {
            if (count.getOrDefault(s.charAt(i), 0) == 0) {
                result = Math.max(result, i - left + 1);
            }
            count.put(s.charAt(i), count.getOrDefault(s.charAt(i), 0) + 1);
            while (count.get(s.charAt(i)) > 1) {
                count.put(s.charAt(left), count.get(s.charAt(left)) - 1);
                left++;
            }
        }

        return result;
    }
}

File: core/tmp/Python/array/longest_substring_without_repeating_characters.py
import collections


def lengthOfLongestSubstring(s: str) -> int:
    count = collections.Counter()
    left, result = 0, 0,
    for index, ch in enumerate(s):
        if count[ch] == 0:
            result = max(result, index - left + 1)
        count[ch] += 1

        while count[ch] > 1:
            count[s[left]] -= 1
            left += 1

    return result


if __name__ == "__main__":
    print(lengthOfLongestSubstring(""))
